src/NLog/Internal/UrlHelper.cs

```diff
@@ -33,82 +33,152 @@
+            { '0', '1', '2', '3', '4', '5', '6', '7',
+            '8', '9', 'a', 'b', 'c', 'd', 'e', 'f' };
+
+        [Flags]
+        public enum EscapeEncodingFlag
+        {
+            /// <summary>Use UnreservedMarks instead of ReservedMarks, as specified by chosen RFC</summary>
+            UriString = 1,
+            /// <summary>Use RFC2396 standard (instead of RFC3986)</summary>
+            LegacyRfc2396 = 2,
+            /// <summary>Should use lowercase when doing HEX escaping of special characters</summary>
+            LowerCaseHex = 4,
+            /// <summary>Replace space ' ' with '+' instead of '%20'</summary>
+            SpaceAsPlus = 8,
+            /// <summary>Skip UTF8 encoding, and prefix special characters with '%u'</summary>
+            NLogLegacy = 16 | LegacyRfc2396 | LowerCaseHex | UriString,
+        };
+
+        private const string RFC2396ReservedMarks = @";/?:@&=+$,";
+        private const string RFC3986ReservedMarks = @":/?#[]@!$&'()*+,;=";
+        private const string RFC2396UnreservedMarks = @"-_.!~*'()";
+        private const string RFC3986UnreservedMarks = @"-._~";
 
         /// <summary>
-        /// Url encode and URL
+        /// Url encode an URL
         /// </summary>
         /// <param name="str">URL to be encoded</param>
         /// <param name="spaceAsPlus">space as + or %20? <c>false</c> (%20) is the safe option.</param>
         /// <returns>Encoded url.</returns>
-        internal static string UrlEncode(string str, bool spaceAsPlus)
+        public static string UrlEncode(string str, bool spaceAsPlus)
         {
-            if (str == null) return string.Empty;
+            if (string.IsNullOrEmpty(str))
+                return string.Empty;
 
             StringBuilder result = new StringBuilder(str.Length + 20);
-            for (int i = 0; i < str.Length; ++i)
-            {
-                char ch = str[i];
+            result.Append(str);
+            EscapeEncodingFlag escapeFlags = EscapeEncodingFlag.NLogLegacy;
+            if (spaceAsPlus)
+                escapeFlags |= EscapeEncodingFlag.SpaceAsPlus;
+            EscapeDataEncode(result, escapeFlags);
+            return result.ToString();
+        }
 
-                if (ch == ' ' && spaceAsPlus)
-                {
-                    result.Append('+');
-                }
-                else if (IsSafeUrlCharacter(ch))
+        /// <summary>
+        /// Escape unicode string data for use in http-requests
+        /// </summary>
+        /// <param name="target">unicode string-data to be encoded</param>
+        /// <param name="flags"><see cref="EscapeEncodingFlag"/>s for how to perform the encoding</param>
+        public static void EscapeDataEncode(StringBuilder target, EscapeEncodingFlag flags)
+        {
+            if (target.Length == 0)
+                return;
+
+            bool isUriString = ((int)flags & (int)EscapeEncodingFlag.UriString) == (int)EscapeEncodingFlag.UriString;
```

Comment:
I think it's performance wise to create a variable for `(int)flags`?  *maybe also for others)